File: process_vcf.R
library(seqminer)
library(dplyr)
library(tidyverse)

census <- read.csv(file="cancer_gene_census.csv", stringsAsFactors=FALSE)
	  census <- census[!grepl(":-", census$Genome.Location),] #get rid of ranges with no values (":-")
	  coding <- read.csv(file = "ccdsGene.txt",header = FALSE, sep = "\t",stringsAsFactors = FALSE)
	  colnames(coding)[which(names(coding) == "V7")] <- "cdsStart"
	  colnames(coding)[which(names(coding) == "V8")] <- "cdsEnd"
	  coding$V3 = gsub('chr', '', coding$V3)
	  allele <- read.csv(file = "MinorAlleleFreq.txt",header = TRUE, sep = "\t",stringsAsFactors = FALSE)
	  allele <- allele[allele$MAF>0.01,]
	  
	  variants <- list()
	  a<-0
	  
	  files <- list.files(path="~/Desktop/vcf", pattern=".vcf.gz$", full.names=TRUE, recursive=FALSE)
	  
	  for (file in files) {
	    df2 <- data.frame(CHROM=integer(0), POS=integer(0), ID=character(0), REF=character(0), ALT=character(0), QUAL=character(0), FILTER=character(0), GENE=character(0))
	    a <- a + 1
	    tt <-  gsub("\\..*$", "", basename(file))
	    print(file)
	    
	    #CANCER GENE CENUS MATCHING + PASS
	    for (i in census$Genome.Location) {
	      sample <- tabix.read.table(file, i)
	      sample <- sample[sample$FILTER == "PASS",]
	      if(nrow(sample) != 0){
	        
	        position <- i
	        genefound <- census$Gene.Symbol[census$Genome.Location == position]
	        sample$GENE <- genefound
	        df2 <- rbind(df2, sample)
	        
	      }
	    }
	    
	    df2 <- select(df2, CHROM, POS, ID, REF, ALT, FILTER, GENE)
	    df2$CODING <- NA
	    
	    # CHECK FOR CODING GENE's
	    #df2 <- df2 %>%
	    #mutate(CODING = map_chr(
	    #.x = POS,
	    #.f = ~ if_else(
	    #condition = any(.x >= coding$cdsStart & .x <= coding$cdsEnd),
	    #true = "YES",
	    #false = NA_character_
	    #)
	    #))
	    
	    df2$CODING <- ifelse(sapply(seq_along(df2$POS), function(i) {
	      inds <- coding$cdsStart <= df2$POS[i] & coding$cdsEnd >= df2$POS[i]
	      any(inds) & (df2$CHROM[i] == coding$V3[which.max(inds)])
	    }), "YES", NA)
	    
	    df2 <- df2[!is.na(df2$CODING),]
	    
	    
	    
	    #CHECK ALLELIC FREQUENCY (COMMON VARIANTS)
	    
	    df3 <- df2 %>% mutate(
	      temp = paste(CHROM, POS, sep = ":")
	    )
	    
	    df3$COMMON <- NA
	    
	    df3 <- df3 %>%
	      mutate(COMMON= map_chr(
	        .x = temp,
	        .f = ~ if_else(
	          condition = any(.x == allele$Location),
	          true = "YES",
	          false = NA_character_
	        )
	      ))
	    
	    df3$temp <- NULL 
	    df3 <- df3[!is.na(df3$COMMON),]
	    
	    a <- a - 1 
	    variants <- append(variants, list(df3), a)
	    names(variants)[(a+1)] <- tt
	    
	    
	    
	  } #end of analysis
	  

File: process_vcf.py
import pandas as pd
import os
import allel as allel
import tabix as tabix
import collections
import pickle

# import cancer gene census
census = pd.read_csv("/Users/imanthonny/Desktop/mutation/cancer_gene_census.csv",delimiter=",")
census = census[~census["Genome Location"].str.contains(":-")]
census["Genome Location"] = "chr" + census["Genome Location"]

# import coding regions
coding = pd.read_csv("/Users/imanthonny/Desktop/mutation/ccdsGene.txt", delimiter="\t", header=None)
coding.rename(columns={coding.columns[6]:'cdsStart', coding.columns[7]:'cdsEnd'}, inplace=True)
coding[2] = coding[2].replace("chr","", regex=True)

# MAF
MAF = pd.read_csv("/Users/imanthonny/Desktop/mutation/MinorAlleleFreq.txt", delimiter="\t")
MAF = MAF[MAF["MAF"]>0.01]

# loop through each VCF to identify if census genome location present
filtered_vcf = pd.DataFrame()
for root, dirs, files in os.walk("/Users/imanthonny/Desktop/mutation/CCLE_VCFs"):
    for file in files:
        if file.endswith("vcf"):
            print(file)
            path = root + "/" + file
            for i in census["Genome Location"]:
                print(i)
                vcf = allel.vcf_to_dataframe(path, region=i)
                if isinstance(vcf,collections.Sized):  # checks if vcf is NOT blank when imported (has census[Genome Location]region)
                        vcf["Sample"] = file.replace(".vcf", "")
                        vcf["Gene"] = census["Gene Symbol"][census["Genome Location"].str.contains(i)].values[0]
                        # if vcf["FILTER_PASS"].any():  # if FILTER_PASS has any True values. 'if not vcf["FILTER_PASS"].any()' would be used for if FILTER pass has any False values.
                        filtered_vcf = filtered_vcf.append(vcf, ignore_index=True)

filtered_vcf = filtered_vcf.drop(columns=["ALT_2","ALT_3", "QUAL"])

# check for coding genes in filtered vcf's
filtered_vcf_2 = pd.DataFrame()
for v in range(0, len(filtered_vcf["POS"])):
    df3 = coding[(coding["cdsStart"] <= filtered_vcf["POS"][v]) & (coding["cdsEnd"] >= filtered_vcf["POS"][v])]
    if isinstance(df3,collections.Sized):
        chr = filtered_vcf["CHROM"][v].replace("chr","")
        df3.rename(columns={df3.columns[2]:'CHR'}, inplace=True)
        # df3["CHR"] = df3["CHR"].astype(str) # isinstance(df3["CHR"], str), checks if object is a string
        matched_chr = df3["CHR"][df3["CHR"].str.contains(chr)]
        # vcf_final = filtered_vcf.iloc[[v]]
        if len(matched_chr) !=  0:
            filtered_vcf.loc[v, "Coding"] = "YES"
        if len(matched_chr) ==  0:
            filtered_vcf.loc[v,"Coding"] = "NO"
        filtered_vcf_2 = filtered_vcf_2.append(filtered_vcf.iloc[[v]], ignore_index=True)


# isolate coding genes only
vcf_coding_only = filtered_vcf_2[filtered_vcf_2["Coding"]=="YES"]

# check for alleleic frequency (common variants only)

vcf_coding_only["CHROM"] = vcf_coding_only["CHROM"].replace("chr","", regex=True)
combined_coordinate = vcf_coding_only["CHROM"] + ":" + vcf_coding_only["POS"].astype(str)
vcf_coding_only["coordinate"] = combined_coordinate
vcf_coding_only = vcf_coding_only[vcf_coding_only["coordinate"].isin(MAF["Location"])]

final_variants = vcf_coding_only
final_variants["Common"] = "YES"

with open('/Users/imanthonny/Desktop/mutation/final_variants.obj', 'wb') as f:
    pickle.dump(final_variants, f)
